packages/swing-store/src/snapStore.js

```diff
@@ -217,63 +210,117 @@ export function makeSnapStore(
 
   /** @param {string} hash */
   function fullPathFromHash(hash) {
-    return resolve(root, baseNameFromHash(hash));
+    return pathResolve(root, baseNameFromHash(hash));
   }
 
   /** @type { Set<string> } */
   const toDelete = new Set();
 
   /**
+   * Creates a new gzipped snapshot file in the `root` directory
+   * and reports information about the process,
+   * including file size and timing metrics.
+   * Note that timing metrics exclude file open.
+   *
    * @param {(filePath: string) => Promise<void>} saveRaw
    * @returns {Promise<SnapshotInfo>}
    */
   async function save(saveRaw) {
-    return withTempName(async tmpSnapPath => {
-      const { duration: rawSaveSeconds } = await measureSeconds(() =>
-        saveRaw(tmpSnapPath),
-      );
-      const { size: rawByteCount } = await stat(tmpSnapPath);
-      const { result: hash, duration: hashSeconds } = await measureSeconds(() =>
-        fileHash(tmpSnapPath),
-      );
-      toDelete.delete(hash);
-      const baseName = baseNameFromHash(hash);
-      const filePath = resolve(root, baseName);
-      const infoBase = {
-        filePath,
-        hash,
-        rawByteCount,
-        rawSaveSeconds,
-        hashSeconds,
-      };
-      const fileStat = await stat(filePath).catch(e => {
-        if (e.code === 'ENOENT') {
-          return undefined;
+    const cleanup = [];
+    return aggregateTryFinally(
+      async () => {
+        // TODO: Refactor to use tmpFile rather than tmpName.
```

Comment:
Thinking about this, we can't use `tmpFile` here since it's not this process that opens the file, but not worth another CI churn to change the comment.

Reply:
The refactoring I have in mind would be updating the signature of `saveRaw` to accept a file handle or stream rather than a file name, but I understand that may not even be possible.

Reply:
Oh yeah we decompress to the file, then tell xsnap the filename, so it'd work.